lib/binding_rust/lib.rs

```diff
@@ -921,6 +956,340 @@ impl<'a, P> TreePropertyCursor<'a, P> {
+
+        // Compile the query.
+        let ptr = unsafe {
+            ffi::ts_query_new(
+                language.0,
+                bytes.as_ptr() as *const c_char,
+                bytes.len() as u32,
+                &mut error_offset as *mut u32,
+                &mut error_type as *mut ffi::TSQueryError,
+            )
+        };
+
+        // On failure, build an error based on the error code and offset.
+        if ptr.is_null() {
+            let offset = error_offset as usize;
+            return if error_type != ffi::TSQueryError_TSQueryErrorSyntax {
+                let suffix = source.split_at(offset).1;
+                let end_offset = suffix
+                    .find(|c| !char::is_alphanumeric(c) && c != '_' && c != '-')
+                    .unwrap_or(source.len());
+                let name = suffix.split_at(end_offset).0;
+                match error_type {
+                    ffi::TSQueryError_TSQueryErrorNodeType => Err(QueryError::NodeType(name)),
+                    ffi::TSQueryError_TSQueryErrorField => Err(QueryError::Field(name)),
+                    ffi::TSQueryError_TSQueryErrorCapture => Err(QueryError::Capture(name)),
+                    _ => Err(QueryError::Syntax(offset)),
+                }
+            } else {
+                Err(QueryError::Syntax(offset))
+            };
+        }
+
+        let string_count = unsafe { ffi::ts_query_string_count(ptr) };
+        let capture_count = unsafe { ffi::ts_query_capture_count(ptr) };
+        let pattern_count = unsafe { ffi::ts_query_pattern_count(ptr) as usize };
+        let mut result = Query {
+            ptr,
+            capture_names: Vec::with_capacity(capture_count as usize),
+            predicates: Vec::with_capacity(pattern_count),
+        };
+
+        // Build a vector of strings to store the capture names.
+        for i in 0..capture_count {
+            unsafe {
+                let mut length = 0u32;
+                let name =
+                    ffi::ts_query_capture_name_for_id(ptr, i, &mut length as *mut u32) as *const u8;
+                let name = slice::from_raw_parts(name, length as usize);
+                let name = str::from_utf8_unchecked(name);
+                result.capture_names.push(name.to_string());
+            }
+        }
+
+        // Build a vector of strings to represent literal values used in predicates.
+        let string_values = (0..string_count)
+            .map(|i| unsafe {
+                let mut length = 0u32;
+                let value =
+                    ffi::ts_query_string_value_for_id(ptr, i as u32, &mut length as *mut u32)
+                        as *const u8;
+                let value = slice::from_raw_parts(value, length as usize);
+                let value = str::from_utf8_unchecked(value);
+                value.to_string()
+            })
+            .collect::<Vec<_>>();
+
+        // Build a vector of predicates for each pattern.
+        for i in 0..pattern_count {
```

Comment:
This loop reads the structured information about the query's *predicates* and converts it into a Rust `enum` which is used to implement the `eq?` and `match?` predicates that I mentioned in the comments.

There's very similar code in `binding.js` for the WASM binding. In both cases, I use the higher-level language's regex-matching facilities to implement `match?`.

So the plan is, every Tree-sitter binding is going to have a little bit of its own duplicated logic for actually implementing the predicate stuff. All the C library does there is expose the parsed predicates to the caller.

Reply:
@maxbrunsfeld I think this is reasonable, but I'm not sure I have groked the C API yet. How much does a `TSQueryPredicateStep` represent? In  `(eq? @ab hi)`, is each of `eq?` and `@ab` and `hi` a separate "step"? So there can never be a predicate like `(func? @aa (xx yy))` (no nesting), right? Can a singe capture have multiple predicates, and how is that represented then?

Reply:
>  In `(eq? @ab hi)`, is each of `eq?` and `@ab` and `hi` a separate "step"?

Yeah, exactly.

> So there can never be a predicate like `(func? @aa (xx yy))` (no nesting), right?

Yeah, with the current API, nested lists like that are not supported in predicates.

> Can a singe capture have multiple predicates, and how is that represented then?

A single pattern can indeed have multiple predicates. So, for example, you can do:

```clojure
; Match bounds checks like `i < items.length`
((binary_expression
  left: (identifier) @variable
  operator: "||"
  right: (member_expression
    property: (property_identifier) @property))

 (match? @variable "^(i|j|k)")
 (eq? @property "length"))
```

The `ts_query_predicates_for_pattern` function returns the *total* number of steps, and a pointer to the first step of the first patttern. Within this buffer, the two individual patterns `(match? @variable "^(i|j|k")` and `(eq? @property "length")` are separated by a *sentinel* step whose `type` is `TSQueryPredicateStepTypeNone`.

It's a little weird, but I wasn't sure of a simpler way to represent an array of arrays with a C API, especially since *internally* I want to store all of the steps in a single buffer.

Reply:
Thanks, it makes sense now.

Reply:
I made this a bit clearer in the docs in my latest commit.